modules/backup-azure/client.go

```diff
@@ -0,0 +1,192 @@
+	base := path.Join(parts...)
+	return path.Join(a.config.BackupPath, base)
+}
+
+func (a *azureClient) GetObject(ctx context.Context, backupID, key string) ([]byte, error) {
+	objectName := a.makeObjectName(backupID, key)
+
+	if err := ctx.Err(); err != nil {
+		return nil, backup.NewErrContextExpired(errors.Wrapf(err, "get object '%s'", objectName))
+	}
+
+	blobDownloadResponse, err := a.client.DownloadStream(ctx, a.config.Container, objectName, nil)
+	if err != nil {
+		if bloberror.HasCode(err, bloberror.BlobNotFound) {
+			return nil, backup.NewErrNotFound(errors.Wrapf(err, "get object '%s'", objectName))
+		}
+		return nil, backup.NewErrInternal(errors.Wrapf(err, "download stream for object '%s'", objectName))
+	}
+
+	reader := blobDownloadResponse.Body
+	downloadData, err := io.ReadAll(reader)
+	if err != nil {
+		return nil, backup.NewErrInternal(errors.Wrapf(err, "read stream for object '%s'", objectName))
+	}
+	err = reader.Close()
+	if err != nil {
+		return nil, backup.NewErrInternal(errors.Wrapf(err, "close stream for object '%s'", objectName))
+	}
+
+	return downloadData, nil
+}
+
+func (a *azureClient) PutFile(ctx context.Context, backupID, key, srcPath string) error {
+	filePath := path.Join(a.dataPath, srcPath)
+	file, err := os.Open(filePath)
+	if err != nil {
+		return backup.NewErrInternal(errors.Wrapf(err, "open file: %q", filePath))
+	}
+	defer file.Close()
+
+	objectName := a.makeObjectName(backupID, key)
+	_, err = a.client.UploadFile(ctx,
+		a.config.Container,
+		objectName,
+		file,
+		&azblob.UploadFileOptions{
+			Metadata: map[string]*string{"backupid": to.Ptr(backupID)},
+			Tags:     map[string]string{"backupid": backupID},
+		})
+	if err != nil {
+		return backup.NewErrInternal(errors.Wrapf(err, "upload file for object '%s'", objectName))
+	}
+
+	return nil
+}
+
+func (a *azureClient) PutObject(ctx context.Context, backupID, key string, data []byte) error {
+	objectName := a.makeObjectName(backupID, key)
+
+	reader := bytes.NewReader(data)
+	_, err := a.client.UploadStream(ctx,
+		a.config.Container,
+		objectName,
+		reader,
+		&azblob.UploadStreamOptions{
+			Metadata: map[string]*string{"backupid": to.Ptr(backupID)},
+			Tags:     map[string]string{"backupid": backupID},
+		})
+	if err != nil {
+		return backup.NewErrInternal(errors.Wrapf(err, "upload stream for object '%s'", objectName))
+	}
+
+	return nil
+}
+
+func (a *azureClient) Initialize(ctx context.Context, backupID string) error {
+	key := "access-check"
+
+	if err := a.PutObject(ctx, backupID, key, []byte("")); err != nil {
+		return errors.Wrap(err, "failed to access-check Azure backup module")
+	}
+
+	objectName := a.makeObjectName(backupID, key)
+	if _, err := a.client.DeleteBlob(ctx, a.config.Container, objectName, nil); err != nil {
+		return errors.Wrap(err, "failed to remove access-check Azure backup module")
+	}
+
+	return nil
+}
+
+func (a *azureClient) WriteToFile(ctx context.Context, backupID, key, destPath string) error {
```

Comment:
WriteToFile loads object into memory before writing it to the file which might be problematic in case of large files.
To avoiding this we can use `client.DownloadFile` instead to write directly to the file.

Reply:
Good point! I have changed the implementation of `WriteToFile` method to address your comment.